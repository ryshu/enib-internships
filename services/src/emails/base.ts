import path from 'path';

import { email } from './config';

/**
 * @summary Method used to send welcom message to new user
 * @param {string} to Dest user
 * @returns {Promise<string>} Resolve: send message
 */
export function sendWelcom(to: string): Promise<string> {
    return email.send({
        template: path.join(__dirname, 'templates', 'welcome'),
        message: { to },
        locals: {},
    });
}
