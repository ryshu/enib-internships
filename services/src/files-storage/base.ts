import multer from 'multer';
import path from 'path';
import httpStatus from 'http-status-codes';

import { pdfFilter } from './filter';
import { APIError } from '../utils/error';

const baseDir = path.join(__dirname, '..', process.env.BASE_STORAGE_DIR);

export const multerPdfStorage = multer.diskStorage({
    destination(_req: any, _file, cb) {
        cb(null, baseDir);
    },
    filename(_req: any, file, cb) {
        if (file.originalname) {
            cb(
                null,
                file.originalname
                    .replace(/[^a-z0-9]/gi, '_')
                    .toLowerCase()
                    .replace(/__+/g, '-'),
            );
        } else {
            cb(new APIError('ID is required to upload a banner', httpStatus.BAD_REQUEST, 1), '');
        }
    },
});

export const uploadHandler = multer({
    storage: multerPdfStorage,
    fileFilter: pdfFilter,
    limits: {
        fileSize: 4000000, // 4 MB
        fields: 20,
        files: 1,
        parts: 20,
    },
});
