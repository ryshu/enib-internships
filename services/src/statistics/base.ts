export declare interface Statistics {
    internships: {
        total: number;
        suggested: number;
        waiting: number;
        availables: number;
        attributed: number;
        validated: number;
        archived: number;
    };

    students: number;
    mentors: number;

    propositions: number;
}

export declare interface CampaignStatistics {
    internships: {
        total: number;
        availables: number;
        attributed: number;
    };

    students: number;
    mentors: number;

    propositions: number;
    campaign: number;
}

export declare const enum INTERNSHIP_MODE {
    SUGGESTED = 'suggest',
    WAITING = 'waiting',
    AVAILABLE = 'available',
    IN_ATTRIBUTION = 'in_attribution',
    ATTRIBUTED = 'attributed',
    VALIDATED = 'validated',
    ARCHIVED = 'archived',
}

export const InternshipsMode: INTERNSHIP_MODE[] = [
    INTERNSHIP_MODE.SUGGESTED,
    INTERNSHIP_MODE.WAITING,
    INTERNSHIP_MODE.AVAILABLE,
    INTERNSHIP_MODE.ATTRIBUTED,
    INTERNSHIP_MODE.VALIDATED,
    INTERNSHIP_MODE.ARCHIVED,
];

export function isInternshipMode(str: string): str is INTERNSHIP_MODE {
    return InternshipsMode.includes(str as INTERNSHIP_MODE);
}
