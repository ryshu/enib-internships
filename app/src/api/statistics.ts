import request from '@/utils/request';

export function getDefaultStats() {
  return {
    internships: {
      archived: 0,
      attributed: 0,
      availables: 0,
      suggested: 0,
      total: 0,
      validated: 0,
      waiting: 0,
    },
    mentors: 0,
    propositions: 0,
    students: 0,
  };
}

export function getDefaultCampaignStats() {
  return {
    campaign: 1,
    internships: {
      attributed: 0,
      availables: 0,
      total: 0,
    },
    mentors: 0,
    propositions: 0,
    students: 0,
  };
}

export const getGlobalStatistics = () =>
  request({ url: `/statistics`, method: 'get' });

export const getCampaignStatistics = (id: number) =>
  request({ url: `/campaigns/${id}/statistics`, method: 'get' });
