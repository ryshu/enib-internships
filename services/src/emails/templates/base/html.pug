doctype transitional
html
    head
        meta(charset="UTF-8")
        meta(content="width=device-width, initial-scale=1", name="viewport")
        meta(name="x-apple-disable-message-reformatting")
        meta(http-equiv="X-UA-Compatible", content="IE=edge")
        meta(content="telephone=no", name="format-detection")
        title Internship manager - Test
        <!--[if (mso 16)]>
        style(type="text/css") a {text-decoration: none;}
        <![endif]-->
        <!--[if gte mso 9]>
        style sup { font-size: 100% !important; }
        <![endif]-->
        <!--[if !mso]>
        style
            inlude ../build/css/index.css
        link(href="https://fonts.googleapis.com/css?family=Roboto:400,400i,700,700i", rel="stylesheet")
        <![endif]-->

    body(style="padding:0;Margin:0;")
        div(class="es-wrapper-color", style="background-color:#F6F6F6;")
        <!--[if gte mso 9]>
        v:background(xmlns:v="urn:schemas-microsoft-com:vml", fill="t")
            v:fill(type="tile", color="#f6f6f6")
        <![endif]-->
        table(class="es-wrapper", width="100%", cellspacing="0", cellpadding="0", style="mso-table-lspace:0pt;mso-table-rspace:0pt;border-collapse:collapse;border-spacing:0px;padding:0;Margin:0;width:100%;height:100%;background-repeat:repeat;background-position:center top;")
            tr(style="border-collapse:collapse;")
                td(class="st-br", valign="top", style="padding:0;Margin:0;")
                    table(class="es-header", cellspacing="0", cellpadding="0", align="center", style="mso-table-lspace:0pt;mso-table-rspace:0pt;border-collapse:collapse;border-spacing:0px;table-layout:fixed !important;width:100%;background-color:transparent;background-repeat:repeat;background-position:center top;")
                        tr(style="border-collapse:collapse;")
                            td(style="padding:0;Margin:0;background-color:#4A4A4A;", bgcolor="#4A4A4A", align="center")
                                div
                                    table(class="es-header-body" style="mso-table-lspace:0pt;mso-table-rspace:0pt;border-collapse:collapse;border-spacing:0px;background-color:transparent;", width="600", cellspacing="0", cellpadding="0", bgcolor="transparent", align="center")
                                        tr(style="border-collapse:collapse;")
                                            td(align="left", style="Margin:0;padding: 10px 20px;")
                                                table(table width="100%", cellspacing="0", cellpadding="0", style="mso-table-lspace:0pt;mso-table-rspace:0pt;border-collapse:collapse;border-spacing:0px;")
                                                    tr(style="border-collapse:collapse;")
                                                        td(width="560", valign="top", align="center", style="padding:0;Margin:0;")
                                                            table()
                                                                tr(style="border-collapse:collapse;")
                                                                    td(align="center" style="padding:0;Margin:0;")
                                                                        a(target="_blank",  href="https://enib.fr", style="-webkit-text-size-adjust:none;-ms-text-size-adjust:none;mso-line-height-rule:exactly;font-family:roboto, 'helvetica neue', helvetica, arial, sans-serif;font-size:14px;text-decoration:underline;color:#1376C8;")
                                                                            img(src="img/logo_enib.png", alt, style="display:block;border:0;outline:none;text-decoration:none;-ms-interpolation-mode:bicubic;width:166px;height:117px;", width="166")
                        tr(style="border-collapse:collapse;")
                            td(height="65", align="center", style="padding:0;Margin:0;")

                    table(class="es-content", cellspacing="0", cellpadding="0", align="center", style="mso-table-lspace:0pt;mso-table-rspace:0pt;border-collapse:collapse;border-spacing:0px;table-layout:fixed !important;width:100%;")
                        tr(style="border-collapse:collapse;")
                            td(style="padding:0;Margin:0;background-color:transparent;", bgcolor="transparent", align="center")
                                block content

                    table(class="es-footer", cellspacing="0", cellpadding="0", align="center", style="mso-table-lspace:0pt;mso-table-rspace:0pt;border-collapse:collapse;border-spacing:0px;table-layout:fixed !important;width:100%;background-color:#F6F6F6;background-repeat:repeat;background-position:center top;")
                        tr(style="border-collapse:collapse;")
                            td(style="padding:0;Margin:0;background-color:#EA8A00;" background-color="#EA8A00" align="center")
                                table(class="es-footer-body", width="600", cellspacing="0", cellpadding="0", bgcolor="#31cb4b", align="center", style="mso-table-lspace:0pt;mso-table-rspace:0pt;border-collapse:collapse;border-spacing:0px;background-color:transparent;")
                                    tr(style="border-collapse:collapse;")
                                        td(style="Margin:0;padding-top:30px;padding-bottom:30px;padding-left:30px;padding-right:30px;border-radius:0px 0px 10px 10px;background-position:left top;background-color:#EFEFEF;", bgcolor="#efefef", align="left")
                                            table(width="100%", cellspacing="0", cellpadding="0", style="mso-table-lspace:0pt;mso-table-rspace:0pt;border-collapse:collapse;border-spacing:0px;")
                                                tr(style="border-collapse:collapse;")
                                                    td(width="540" align="center" style="padding:0;Margin:0;")
                                                        table(width="100%", cellspacing="0", cellpadding="0", style="mso-table-lspace:0pt;mso-table-rspace:0pt;border-collapse:collapse;border-spacing:0px;")
                                                            tr(style="border-collapse:collapse;")
                                                                td(class="es-m-txt-c" align="center" style="padding:0;Margin:0;")
                                                                    table(class="es-table-not-adapt es-social", cellspacing="0", cellpadding="0", style="mso-table-lspace:0pt;mso-table-rspace:0pt;border-collapse:collapse;border-spacing:0px;")
                                                                        tr(style="border-collapse:collapse;")
                                                                            td(valign="top", align="center", style="padding:0;Margin:0;padding-right:10px;")
                                                                                a(target="_blank", href="https://www.facebook.com/enibrest", style="-webkit-text-size-adjust:none;-ms-text-size-adjust:none;mso-line-height-rule:exactly;font-family:roboto, 'helvetica neue', helvetica, arial, sans-serif;font-size:16px;text-decoration:underline;color:#FFFFFF;")
                                                                                    img(title="Facebook", src="img/facebook-logo-black.png", alt="Fb", width="32", style="display:block;border:0;outline:none;text-decoration:none;-ms-interpolation-mode:bicubic;")
                                                                            td(valign="top", align="center", style="padding:0;Margin:0;padding-right:10px;")
                                                                                a(target="_blank", href="https://twitter.com/enib_info", style="-webkit-text-size-adjust:none;-ms-text-size-adjust:none;mso-line-height-rule:exactly;font-family:roboto, 'helvetica neue', helvetica, arial, sans-serif;font-size:16px;text-decoration:underline;color:#FFFFFF;")
                                                                                    img(title="Twitter", src="img/twitter-logo-black.png", alt="Tw", width="32", style="display:block;border:0;outline:none;text-decoration:none;-ms-interpolation-mode:bicubic;")
                                                                            td(valign="top", align="center", style="padding:0;Margin:0;padding-right:10px;")
                                                                                a(target="_blank", href="https://enib.fr", style="-webkit-text-size-adjust:none;-ms-text-size-adjust:none;mso-line-height-rule:exactly;font-family:roboto, 'helvetica neue', helvetica, arial, sans-serif;font-size:16px;text-decoration:underline;color:#FFFFFF;")
                                                                                    img(title="Website", src="img/link-logo-black.png", alt="Website", width="32", style="display:block;border:0;outline:none;text-decoration:none;-ms-interpolation-mode:bicubic;")
                                                                            td(valign="top", align="center", style="padding:0;Margin:0;padding-right:10px;")
                                                                                a(target="_blank", href="https://ryshu.github.io/enib-internships/", style="-webkit-text-size-adjust:none;-ms-text-size-adjust:none;mso-line-height-rule:exactly;font-family:roboto, 'helvetica neue', helvetica, arial, sans-serif;font-size:16px;text-decoration:underline;color:#FFFFFF;")
                                                                                    img(title="GitHub", src="img/github-logo-black.png", alt="GitHub", width="32", style="display:block;border:0;outline:none;text-decoration:none;-ms-interpolation-mode:bicubic;")
                                                                            td(valign="top", align="center", style="padding:0;Margin:0;")
                                                                                a(target="_blank", href="https://www.enib.fr/fr/contact-fr", style="-webkit-text-size-adjust:none;-ms-text-size-adjust:none;mso-line-height-rule:exactly;font-family:roboto, 'helvetica neue', helvetica, arial, sans-serif;font-size:16px;text-decoration:underline;color:#FFFFFF;")
                                                                                    img(title="Email", src="img/mail-logo-black.png", alt="Email", width="32", style="display:block;border:0;outline:none;text-decoration:none;-ms-interpolation-mode:bicubic;")

                                    tr(style="border-collapse:collapse;")
                                        td(style="padding:0;Margin:0;background-position:left top;", align="left")
                                            table(width="100%", cellspacing="0", cellpadding="0", style="mso-table-lspace:0pt;mso-table-rspace:0pt;border-collapse:collapse;border-spacing:0px;")
                                                tr(style="border-collapse:collapse;")
                                                    td(width="600", valign="top", align="center", style="padding:0;Margin:0;")
                                                        table(width="100%", cellspacing="0", cellpadding="0", style="mso-table-lspace:0pt;mso-table-rspace:0pt;border-collapse:collapse;border-spacing:0px;")
                                                            tr(style="border-collapse:collapse;")
                                                                td(height="40", align="center", style="padding:0;Margin:0;")
