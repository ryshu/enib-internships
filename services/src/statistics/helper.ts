import { cloneDeep } from 'lodash';

import { Statistics, CampaignStatistics } from './base';

export function getCleanStatistics(stats: Partial<Statistics>): Statistics {
    const tmp = cloneDeep(stats);
    return {
        internships: tmp.internships
            ? {
                  abroad: tmp.internships.abroad || 0,
                  total: tmp.internships.total || 0,
                  waiting: tmp.internships.waiting || 0,
                  published: tmp.internships.published || 0,
                  attributed_student: tmp.internships.attributed_student || 0,
                  attributed_mentor: tmp.internships.attributed_mentor || 0,
                  available_campaign: tmp.internships.available_campaign || 0,
                  running: tmp.internships.running || 0,
                  validation: tmp.internships.validation || 0,
                  archived: tmp.internships.archived || 0,
              }
            : {
                  abroad: 0,
                  total: 0,
                  waiting: 0,
                  published: 0,
                  attributed_student: 0,
                  attributed_mentor: 0,
                  available_campaign: 0,
                  running: 0,
                  validation: 0,
                  archived: 0,
              },

        mentors: tmp.mentors || 0,
        students: tmp.students || 0,
        propositions: tmp.propositions || 0,
    };
}

export function getCleanCampaignStatistics(stats: Partial<CampaignStatistics>): CampaignStatistics {
    const tmp = cloneDeep(stats);
    return {
        internships: tmp.internships
            ? {
                  total: tmp.internships.total || 0,
                  availables: tmp.internships.availables || 0,
                  attributed: tmp.internships.attributed || 0,
              }
            : { total: 0, availables: 0, attributed: 0 },

        mentors: tmp.mentors || 0,
        students: tmp.students || 0,
        propositions: tmp.propositions || 0,
        campaign: tmp.campaign,
    };
}
