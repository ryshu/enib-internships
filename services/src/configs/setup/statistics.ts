import { Op } from 'sequelize';

import logger from '../../utils/logger';

import cache from '../../statistics/singleton';

import Internships from '../../models/Internships';
import Students from '../../models/Students';
import Mentors from '../../models/Mentors';
import MentoringPropositions from '../../models/MentoringPropositions';
import Campaigns from '../../models/Campaigns';
import { CampaignStatistics } from '../../statistics/base';

export async function setupStatistics() {
    async function _searchCampaign(c: Campaigns): Promise<CampaignStatistics> {
        const campaignCond = {
            [Op.or]: [{ validatedCampaignId: c.id }, { availableCampaignId: c.id }],
        };
        return {
            internships: {
                total: await Internships.count({ where: campaignCond }),
                availables: await Internships.count({ where: { availableCampaignId: c.id } }),
                attributed: await Internships.count({ where: { validatedCampaignId: c.id } }),
            },
            students: await Internships.count({
                where: { ...campaignCond, [Op.not]: { studentId: null } },
            }),
            mentors: await c.countMentors(),

            propositions: await c.countPropositions(),
            campaign: c.id,
        };
    }
    const global = {
        internships: {
            total: await Internships.count(),
            suggested: await Internships.count({ where: { state: 'suggest' } }),
            waiting: await Internships.count({ where: { state: 'waiting' } }),
            availables: await Internships.count({ where: { state: 'available' } }),
            validated: await Internships.count({ where: { state: 'validated' } }),
            attributed: await Internships.count({ where: { state: 'attributed' } }),
            archived: await Internships.count({ where: { state: 'archived' } }),
        },
        students: await Students.count(),
        mentors: await Mentors.count(),

        propositions: await MentoringPropositions.count(),
    };

    const campaigns = await Campaigns.findAll();
    const stats = await Promise.all(campaigns.map((c) => _searchCampaign(c)));

    cache.reset();
    cache.init(global, ...stats);
    logger.info(`STATISTICS - ${JSON.stringify(global, null, 4)}`);
    stats.forEach((s) => logger.info(`STATISTICS - ${JSON.stringify(s, null, 4)}`));
}
