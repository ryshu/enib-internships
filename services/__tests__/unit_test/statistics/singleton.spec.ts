import cache from '../../../src/statistics/singleton';
import { Statistics, INTERNSHIP_MODE } from '../../../src/statistics/base';

const DEFAULT_STATISTICS: Statistics = {
    internships: {
        total: 250,
        suggested: 10,
        waiting: 30,
        availables: 150,
        attributed: 50,
        validated: 20,
        archived: 0,
    },
    students: 150,
    mentors: 200,
    propositions: 300,
};

const DEFAULT_CAMPAIGN_STATISTICS: any = [
    {
        internships: { total: 250, availables: 150, attributed: 100 },
        students: 150,
        mentors: 200,
        propositions: 300,
    },
    {
        internships: { total: 250, availables: 150, attributed: 100 },
        students: 150,
        mentors: 200,
        propositions: 300,
        campaign: 10,
    },
];

describe('StatisticsCache', () => {
    beforeEach((done) => {
        // '#.reset()'
        cache.reset();

        // '#.init()'
        cache.init(DEFAULT_STATISTICS, ...DEFAULT_CAMPAIGN_STATISTICS);

        done();
    });

    it('#.reset()', () => {
        cache.reset();
        expect(cache.initialized).toBeFalsy();
        cache.reset();
    });

    it('#.init()', () => {
        cache.reset();
        expect(cache.initialized).toBeFalsy();

        cache.init(DEFAULT_STATISTICS, ...DEFAULT_CAMPAIGN_STATISTICS);
        expect(cache.statistics).toEqual(DEFAULT_STATISTICS);
        expect(cache.getCampaign(10)).toEqual(DEFAULT_CAMPAIGN_STATISTICS[1]);
        cache.init(DEFAULT_STATISTICS, ...DEFAULT_CAMPAIGN_STATISTICS);
    });

    it('#.stateChange ARCHIVED -> ATTRIBUTED', () => {
        cache.stateChange(INTERNSHIP_MODE.ARCHIVED, INTERNSHIP_MODE.ATTRIBUTED, 10);
        expect(cache.getCampaign(10)).toMatchSnapshot();
    });

    it('#.stateChange AVAILABLE -> VALIDATED', () => {
        cache.stateChange(INTERNSHIP_MODE.AVAILABLE, INTERNSHIP_MODE.VALIDATED, 10);
        expect(cache.getCampaign(10)).toMatchSnapshot();
    });

    it('#.stateChange AVAILABLE -> WAITING', () => {
        cache.stateChange(INTERNSHIP_MODE.AVAILABLE, INTERNSHIP_MODE.WAITING, 10);
        expect(cache.getCampaign(10)).toMatchSnapshot();
    });

    it('#.stateChange INVALID -> WAITING', () => {
        cache.stateChange('INVALID' as any, INTERNSHIP_MODE.WAITING, 10);
        expect(cache.getCampaign(10)).toMatchSnapshot();
    });

    it('#.stateAdd INVALID', () => {
        cache.stateAdd('INVALID' as any, 10);
        expect(cache.getCampaign(10)).toMatchSnapshot();
    });

    it('#.stateRemove INVALID', () => {
        cache.stateRemove('INVALID' as any, 10);
        expect(cache.getCampaign(10)).toMatchSnapshot();
    });

    it('#.addMentor', () => {
        // Default + campaign
        cache.addMentor();
        cache.addMentor();
        cache.addMentor();

        expect(cache.statistics.mentors).toEqual(DEFAULT_STATISTICS.mentors + 3);
    });

    it('#.removeMentor', () => {
        // Default + campaign
        cache.removeMentor();
        cache.removeMentor();
        cache.removeMentor();

        expect(cache.statistics.mentors).toEqual(DEFAULT_STATISTICS.mentors - 3);
    });

    it('#.linkMentor', () => {
        // Default + campaign
        cache.linkMentor(10);
        expect(cache.getCampaign(10)!.mentors).toEqual(DEFAULT_CAMPAIGN_STATISTICS[1].mentors + 1);

        // Default + new campaign
        cache.linkMentor(500);
        expect(cache.getCampaign(500)).toEqual({
            internships: { total: 0, availables: 0, attributed: 0 },
            students: 0,
            mentors: 1,
            propositions: 0,
            campaign: 500,
        });

        cache.linkMentor(null as any);
    });

    it('#.unlinkMentor', () => {
        // Default + campaign
        cache.unlinkMentor(10);
        expect(cache.getCampaign(10)!.mentors).toEqual(DEFAULT_CAMPAIGN_STATISTICS[1].mentors - 1);

        // Default + new campaign
        cache.unlinkMentor(500);
        expect(cache.getCampaign(500)).toEqual({
            internships: { total: 0, availables: 0, attributed: 0 },
            students: 0,
            mentors: -1,
            propositions: 0,
            campaign: 500,
        });

        cache.unlinkMentor(null as any);
    });

    it('#.addStudent', () => {
        // Only default statistics
        cache.addStudent();

        expect(cache.statistics.students).toEqual(DEFAULT_STATISTICS.students + 1);

        // Default + campaign
        cache.addStudent(10);

        expect(cache.statistics.students).toEqual(DEFAULT_STATISTICS.students + 2);
        expect(cache.getCampaign(10)!.students).toEqual(
            DEFAULT_CAMPAIGN_STATISTICS[1].students + 1,
        );

        // Default + new campaign
        cache.addStudent(453);
        expect(cache.getCampaign(453)).toEqual({
            internships: { total: 0, availables: 0, attributed: 0 },
            students: 1,
            mentors: 0,
            propositions: 0,
            campaign: 453,
        });
    });

    it('#.removeStudent', () => {
        // Only default statistics
        cache.removeStudent();

        expect(cache.statistics.students).toEqual(DEFAULT_STATISTICS.students - 1);

        // Default + campaign
        cache.removeStudent(10);

        expect(cache.statistics.students).toEqual(DEFAULT_STATISTICS.students - 2);
        expect(cache.getCampaign(10)!.students).toEqual(
            DEFAULT_CAMPAIGN_STATISTICS[1].students - 1,
        );

        // Default + new campaign
        cache.removeStudent(454);
        expect(cache.getCampaign(454)).toBeFalsy();
    });

    it('#.linkStudent', () => {
        // Default + campaign
        cache.linkStudent(10);
        expect(cache.getCampaign(10)!.students).toEqual(
            DEFAULT_CAMPAIGN_STATISTICS[1].students + 1,
        );

        // Default + new campaign
        cache.linkStudent(501);
        expect(cache.getCampaign(501)).toEqual({
            internships: { total: 0, availables: 0, attributed: 0 },
            students: 1,
            mentors: 0,
            propositions: 0,
            campaign: 501,
        });

        cache.linkStudent(null as any);
    });

    it('#.addProposition', () => {
        // Only default statistics
        cache.addProposition();

        expect(cache.statistics.propositions).toEqual(DEFAULT_STATISTICS.propositions + 1);

        // Default + campaign
        cache.addProposition(10);

        expect(cache.statistics.propositions).toEqual(DEFAULT_STATISTICS.propositions + 2);
        expect(cache.getCampaign(10)!.propositions).toEqual(
            DEFAULT_CAMPAIGN_STATISTICS[1].propositions + 1,
        );

        // Default + new campaign
        cache.addProposition(200);
        expect(cache.getCampaign(200)).toEqual({
            internships: { total: 0, availables: 0, attributed: 0 },
            students: 0,
            mentors: 0,
            propositions: 1,
            campaign: 200,
        });
    });

    it('#.removeProposition', () => {
        // Only default statistics
        cache.removeProposition();

        expect(cache.statistics.propositions).toEqual(DEFAULT_STATISTICS.propositions - 1);

        // Default + campaign
        cache.removeProposition(10);

        expect(cache.statistics.propositions).toEqual(DEFAULT_STATISTICS.propositions - 2);
        expect(cache.getCampaign(10)!.propositions).toEqual(
            DEFAULT_CAMPAIGN_STATISTICS[1].propositions - 1,
        );

        // Default + new campaign
        cache.removeProposition(201);
        expect(cache.getCampaign(201)).toBeFalsy();
    });

    it('#.linkProposition', () => {
        // Default + campaign
        cache.linkProposition(10);
        expect(cache.getCampaign(10)!.propositions).toEqual(
            DEFAULT_CAMPAIGN_STATISTICS[1].propositions + 1,
        );

        // Default + new campaign
        cache.linkProposition(202);
        expect(cache.getCampaign(202)).toEqual({
            internships: { total: 0, availables: 0, attributed: 0 },
            students: 0,
            mentors: 0,
            propositions: 1,
            campaign: 202,
        });

        cache.linkProposition(null as any);
    });

    it('#.getCampaign', () => {
        expect(cache.getCampaign(10)).toBeTruthy();
        expect(cache.getCampaign(454)).toBeFalsy();
    });

    it('#.newCampain', () => {
        expect(cache.getCampaign(80)).toBeFalsy();
        cache.newCampain(80, {});
        expect(cache.getCampaign(80)).toBeTruthy();
    });

    it('#.isDefined', () => {
        expect(cache.isDefined(10)).toBeTruthy();
        expect(cache.isDefined(454)).toBeFalsy();
    });
});
