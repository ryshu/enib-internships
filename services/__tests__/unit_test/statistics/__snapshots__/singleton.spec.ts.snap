// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatisticsCache #.stateAdd INVALID 1`] = `
Object {
  "campaign": 10,
  "internships": Object {
    "attributed": 100,
    "availables": 150,
    "total": 250,
  },
  "mentors": 200,
  "propositions": 300,
  "students": 150,
}
`;

exports[`StatisticsCache #.stateChange ARCHIVED -> ATTRIBUTED 1`] = `
Object {
  "campaign": 10,
  "internships": Object {
    "attributed": 99,
    "availables": 150,
    "total": 250,
  },
  "mentors": 200,
  "propositions": 300,
  "students": 150,
}
`;

exports[`StatisticsCache #.stateChange AVAILABLE -> VALIDATED 1`] = `
Object {
  "campaign": 10,
  "internships": Object {
    "attributed": 100,
    "availables": 151,
    "total": 250,
  },
  "mentors": 200,
  "propositions": 300,
  "students": 150,
}
`;

exports[`StatisticsCache #.stateChange AVAILABLE -> WAITING 1`] = `
Object {
  "campaign": 10,
  "internships": Object {
    "attributed": 100,
    "availables": 151,
    "total": 250,
  },
  "mentors": 200,
  "propositions": 300,
  "students": 150,
}
`;

exports[`StatisticsCache #.stateChange INVALID -> WAITING 1`] = `
Object {
  "campaign": 10,
  "internships": Object {
    "attributed": 100,
    "availables": 150,
    "total": 250,
  },
  "mentors": 200,
  "propositions": 300,
  "students": 150,
}
`;

exports[`StatisticsCache #.stateRemove INVALID 1`] = `
Object {
  "campaign": 10,
  "internships": Object {
    "attributed": 100,
    "availables": 150,
    "total": 250,
  },
  "mentors": 200,
  "propositions": 300,
  "students": 150,
}
`;
